backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java

```diff
@@ -237,6 +238,117 @@ public void revokePermission(UUID bucketId, UUID userId) {
+            .orElseThrow(AccessTypeNotFoundException::new);
+
+    FileMetadataEntity file =
+        fileMetadataRepository
+            .findById(permissionGrantRequest.getObjectId())
+            .orElseThrow(FileNotFoundException::new);
+
+    if (file.isDeleted()) {
+      throw new FileAlreadyDeletedException();
+    }
+
+    permission.setGrantedBy(currentUser);
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessType);
+    permission.setObject(file);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    currentUserHasPermissionForFile(file, AccessType.OWNER);
+
+    if (permissionRepository.existsById(
+        new UserObjectAccessKey(targetUser.getId(), file.getId()))) {
+      throw new PermissionAlreadyExistsException();
+    }
+
+    UserObjectAccessEntity savedPermission = permissionRepository.save(permission);
+
+    grantViewPermissionUpstream(file, currentUser, targetUser, permissionGrantRequest);
+
+    log.info(
+        "User {} ({}) granted {} permission for file {} ({}) to user {} ({})",
+        permission.getGrantedBy().getEmail(),
+        permission.getGrantedBy().getId(),
+        permission.getAccess().getName(),
+        permission.getObject().getName(),
+        permission.getObject().getId(),
+        permission.getGrantedTo().getEmail(),
+        permission.getGrantedTo().getId());
+
+    return modelMapper.map(savedPermission, PermissionResponse.class);
+  }
+
+  private void grantViewPermissionUpstream(
+      FileMetadataEntity file,
+      UserEntity currentUser,
+      UserEntity targetUser,
+      PermissionGrantRequest permissionGrantRequest) {
+
+    grantViewPermissionToAllFoldersUpstream(file.getParentFolder(), currentUser, targetUser);
+    permissionGrantRequest.setObjectId(file.getBucket().getId());
+    permissionGrantRequest.setAccess("view");
+    try {
+      grantPermissionForBucket(permissionGrantRequest);
+    } catch (SkyNestBaseException ignored) {
+    }
+  }
+
+  private void grantViewPermissionToAllFoldersUpstream(
+      FolderEntity parent, UserEntity currentUser, UserEntity targetUser) {
+
+    if (parent == null
+        || permissionRepository.existsById(
+            new UserObjectAccessKey(targetUser.getId(), parent.getId()))) {
+      return;
+    }
+
+    AccessTypeEntity accessTypeView =
+        accessTypeRepository
+            .findByName(AccessType.VIEW.getText())
+            .orElseThrow(AccessTypeNotFoundException::new);
+
+    UserObjectAccessEntity permission = new UserObjectAccessEntity();
+    permission.setGrantedBy(currentUser);
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessTypeView);
+    permission.setObject(parent);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    permissionRepository.save(permission);
```

Comment:
what if the current user doesn't have permission to do this?
should probably be fine though

Reply:
I'm assuming this method will only be called from inside this service, after the permission check